### ISSUE TITLE
Emit JS

### ISSUE THREAD

**Issue by 22459** [PiotrDabkowski/Js2Py#97]
I'm not sure an issue is really an appropriate place for this, but i didn't have any better ideas; 
is there an easy way to emit an AST obtained from js2py.parse_js as javascript code again?

Thanks.

**Comment by 22459** [comment]
Actually, I've looked at several python javascript parsing libraries now, among related things, so I might be getting lost, but I think I read that Js2Py also uses some variation of esprima, like https://github.com/PiotrDabkowski/pyjsparser , ....Oh wow, I only now noticed that you also made pyjsparser. So to rephrase, is pyjsparser compatible with the output of parse_js()?

**Comment by PiotrDabkowski** [comment]
Yes, the outputs should be exactly the same, js2py imports pyjsparser, both emit esprima AST.

Js2Py cannot convert AST to the JacaScript code, only to the Python code. But maybe you can find a tool that does it.

**Comment by PiotrDabkowski** [comment]
By the way, if you are using Python 2 then you can try the newest Js2Py that supports requiring npm modules. Node has one module called `escodegen` that does exactly what you want

```
from js2py import require, parse_js
escodegen = require('escodegen')
escodegen.generate( parse_js("var hello = 'world!' ") )
```

**Comment by 22459** [comment]
Ah, I'll look into that, thanks. I haven't had time to make any progress with this, perhaps in the next few days or over the weekend.

However, first I need to find a nice way to walk the AST anyway.
Do you have any suggestions?
From what I've read the visitor pattern is the usual algorithm to use here, and it's certainly doable, but I haven't been able to think of a **nice** way to do it. The data structure seems to be a bit inhomogeneous; there are dicts and there are lists and the children of the dicts depend on a type specifying key.

**Comment by PiotrDabkowski** [comment]
This is as far as I know the best approach, this is the exact way I use to translate AST to python, see [here](https://github.com/PiotrDabkowski/Js2Py/blob/0f3d4bde0a5e6e646803a5ae95542e0068c9087c/js2py/translators/translating_nodes.py#L112). It is very easy to understand and as there is one function for each code syntax element that handles all the details. There is quite a few functions but they are very easy to write and in total they add up to about 500 lines so thats not bad.

**Comment by 22459** [comment]
Ok thanks. With any luck I can probably just reuse the one from slimit and/or yours as well.

**Comment by 22459** [comment]
Still haven't gotten around to coding anything myself but I ran across this, it may be convenient for anyone that runs across this thread:
[https://github.com/austinbyers/esprima-ast-visitor ](https://github.com/austinbyers/esprima-ast-visitor )

**Comment by 22459** [comment]
In the end my solution to this is to use https://pypi.python.org/pypi/python-bond/1.4 to communicate with node.js and generate javascript code by passing through the JSON encoded AST to escodegen.
